_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+  // ...
+```
+
+As you can see, the repetition of `executor` quickly becomes a tedious task.
+We can declare that this parameter is common to the current context and avoid its repetition:
+
+```scala
+@main def main =
+  given executor as ExecutionContext = ExecutionContext.global
+  val fact100 = Future(factorial(100))
+  val fibo100 = Future(fibonacci(100))
+```
+
+This works because the `apply` method of future could be declared as follows:
+
+```scala
+// scala.concurrent.Future.scala
+object Future {
+    // ..
+    def apply[T](body: => T)(using executor: ExecutionContext): Future[T] = // ...
+}
+```
+We see here the other half of the syntax: if we declare a parameter as `using`, the compiler
+will search the current scope at call sites for `given` values with a compatible type.
+
+Note that the `executor` identifier is never used, so we can omit it:
+
+```scala
+given ExecutionContext = ExecutionContext.global
+```
+the same is possible for the `using` parameter:
+```scala
+def apply[T](body: => T)(using ExecutionContext): Future[T] = // ...
+```
+Note that in Scala 3 the type of a `given` definition must be explicit.
+
+Another important aspect is how we import these values. For example, if you had a
+more involved definition for the `ExecutionContext` that is used in multiple
+files, you could refactor it into a different file:
+
+```scala
+// Context.scala
+import scala.concurrent.ExecutionContext
+import java.util.concurrent.Executors
+
+object Context:
+  given ExecutionContext =
+    ExecutionContext.fromExecutor(Executors.newFixedThreadPool(5))
+```
+
+Then you can import it using a wildcard:
+
+```scala
+import Context.given
+```
+Or by making the type that you are bringing in scope explicit:
+```
+Import Context.{given ExecutionContext}
+```
+This allows you to have more control over imports without relying on instance names.
+
+### Find out more
+
+You can learn more about `using` / `given` and about the rules of resolution in
+[the documentation](https://dotty.epfl.ch/docs/reference/contextual/motivation.html).
+
+### How was this done in Scala 2?
+
+In Scala 2 this pattern was achieved by marking both the value and the parameter with the
+`implicit` keyword. The previous example would look like this:
+
+```scala
+// Main.scala
+import scala.concurrent._
+object Main extends App
+  implicit val ec: ExecutionContext = ExecutionContext.global
+  Future(println("Hello World"))
+}
+
+// scala.concurrent.Future.scala
+object Future {
+    // ..
+    def apply[T](body: => T)(implicit executor: ExecutionContext): Future[T] = // ...
+}
+```
+
+We note again that we had to provide a name which might not be used for the variable. The `implictly` function from Scala 2 is renamed to `summon` in Scala 3.
+
+Finally the special import syntax allows users to explicitly import `given` instances (either by wildcard or by type) and not to have undesired instances obtained by a vague `import Context._`
```

Comment:
The stronger added value of `import given` is that we can import by type. `import Context._` might be vague, but so is `import Foo.given`, and yet that still exists. I would put more emphasis on the type aspect, for example:

> Finally, the special import syntax allows users to import `given` instances by type, rather than by name. This makes more sense since we usually refer to them by type as well.

Reply:
Your formulation is clearer than mine, thanks